src/Xamarin.Android.Build.Tasks/Utilities/TypeMappingDebugNativeAssemblyGenerator.cs

```diff
@@ -28,115 +81,98 @@ public TypeMappingDebugNativeAssemblyGenerator (NativeAssemblerTargetProvider ta
+			string javaToManagedSymbolName;
+			generator.WriteCommentLine ("Java to managed map: START");
+			generator.WriteDataSection ($"rel.{JavaToManagedSymbol}");
 
-			size = WriteStructure (output, packed: false, structureWriter: () => WriteTypeMapStruct (output));
-
-			WriteStructureSize (output, TypeMapSymbol, size);
-		}
-
-		// MUST match the TypeMap struct from src/monodroid/xamarin-app.hh
-		uint WriteTypeMapStruct (StreamWriter output)
-		{
-			uint size = 0;
+			if (haveJavaToManaged) {
+				NativeAssemblyGenerator.StructureWriteContext mapArray = generator.StartStructureArray ();
+				var map_entry = new TypeMapEntry ();
+				foreach (TypeMapGenerator.TypeMapDebugEntry entry in data.JavaToManagedMap) {
+					map_entry.from = entry.JavaLabel;
 
-			WriteCommentLine (output, "entry_count");
-			size += WriteData (output, data.EntryCount);
+					TypeMapGenerator.TypeMapDebugEntry managedEntry = entry.DuplicateForJavaToManaged != null ? entry.DuplicateForJavaToManaged : entry;
+					map_entry.to = managedEntry.SkipInJavaToManaged ? null : managedEntry.ManagedLabel;
 
-			WriteCommentLine (output, "assembly_name (unused in this mode)");
-			size += WritePointer (output);
+					NativeAssemblyGenerator.StructureWriteContext mapEntryStruct = generator.AddStructureArrayElement (mapArray);
+					generator.WriteStructure (mapEntryStruct, map_entry);
+				}
 
-			WriteCommentLine (output, "data (unused in this mode)");
-			size += WritePointer (output);
+				javaToManagedSymbolName = generator.WriteSymbol (mapArray, JavaToManagedSymbol);
+			} else {
+				javaToManagedSymbolName = generator.WriteEmptySymbol (SymbolType.Object, JavaToManagedSymbol);
+			}
+			generator.WriteCommentLine ("Java to managed map: END");
 
-			WriteCommentLine (output, "java_to_managed");
-			size += WritePointer (output, JavaToManagedSymbol);
+			generator.WriteCommentLine ("TypeMap structure");
+			generator.WriteDataSection ($"rel.ro.{TypeMapSymbol}");
 
-			WriteCommentLine (output, "managed_to_java");
-			size += WritePointer (output, ManagedToJavaSymbol);
+			var type_map = new TypeMap {
+				entry_count = (uint)data.EntryCount,
+				java_to_managed = javaToManagedSymbolName,
+				managed_to_java = managedToJavaSymbolName,
+			};
 
-			return size;
+			NativeAssemblyGenerator.StructureWriteContext typeMapStruct = generator.StartStructure ();
+			generator.WriteStructure (typeMapStruct, type_map);
+			generator.WriteSymbol (typeMapStruct, TypeMapSymbol, local: false);
 		}
 
-		void WriteSharedBits (StreamWriter output, bool haveJavaToManaged, bool haveManagedToJava)
+		void WriteSharedBits (NativeAssemblyGenerator generator, StreamWriter output, bool haveJavaToManaged, bool haveManagedToJava)
 		{
-			string label;
-
 			if (haveJavaToManaged) {
-				WriteCommentLine (output, "Java type names: START");
+				generator.WriteStringSection (output, "java_type_names");
+				generator.WriteCommentLine (output, "Java type names: START", useBlockComment: true);
```

Comment:
Why does this block need to use "block comments" instead of line comments?  This method appears to be the only use of `useBlockComment: true`.

I also don't understand how you *terminate* a block comment.

Reply:
Part of why I ask is that *support* for "block comments" appears to go *everywhere*, and it feels like things would get much simpler if you removed support for block comments, given that it's really only used in one place.

Reply:
Block comments look better sometimes - I care about that when I have to read the generated code, I want the code to be readable and nicely formatted. And the fact that they aren't used everywhere right now, doesn't mean they won't be used somewhere in the future when we add some code.